src/main/java/com/conveyal/taui/models/ModificationStop.java

```diff
@@ -30,177 +28,215 @@
     private final boolean autoGenerated;
 
     /**
      * Constructor for ModificationStops. It creates an immutable object with all fields final.
      */
-    private ModificationStop (Coordinate c, String id, double distanceFromStart, boolean autoGenerated) {
-        this.coordinate = c;
-        this.id = id;
-        this.distanceFromStart = distanceFromStart;
+    private ModificationStop (StopSpec stopSpec, int dwellTimeSeconds, int hopTimeSeconds, boolean autoGenerated) {
+        this.stopSpec = stopSpec;
+        this.dwellTimeSeconds = dwellTimeSeconds;
+        this.hopTimeSeconds = hopTimeSeconds;
         this.autoGenerated = autoGenerated;
     }
 
-    /**
-     * Convert a list of ModificationStops (which are internal to the backend conversion process) to a list of the
-     * StopSpec type required by r5.
-     */
-    static List<StopSpec> toStopSpecs (List<ModificationStop> stops) {
-        return stops.stream()
-                .map(s -> {
-                    if (s.id == null || s.autoGenerated){
-                        return new StopSpec(s.coordinate.x, s.coordinate.y);
-                    } else {
-                        // Stop references an existing GTFS stop ID, send that instead of its coordinates.
-                        return new StopSpec(s.id);
-                    }
-                })
-                .collect(Collectors.toList());
-    }
-
     /**
      * Convert a list of Segments from a modification (the front-end representation) to this internal
      * backend representation.
      */
-    static List<ModificationStop> getStopsFromSegments (List<Segment> segments) {
+    static List<ModificationStop> getStopsFromSegments (List<Segment> segments, Integer[] dwellTimes, int defaultDwellTime, int[] segmentSpeedsKph) {
+        if (segments == null || segments.size() == 0) {
+            return new ArrayList<>();
+        }
 
         // Keep a stack of Stops because as part of auto-generating stops we sometimes need to back one out.
         Stack<ModificationStop> stops = new Stack<>();
-        CoordinateReferenceSystem crs = DefaultGeographicCRS.WGS84;
 
-        if (segments == null || segments.size() == 0) {
-            return new ArrayList<>();
+        // Create an empty array if dwell times are null
+        if (dwellTimes == null) {
+            dwellTimes = new Integer[0];
         }
 
         Segment firstSegment = segments.get(0);
+        int realStopIndex = 0;
 
         if (firstSegment.stopAtStart) {
-            stops.add(new ModificationStop(firstSegment.geometry.getCoordinates()[0],
-                    firstSegment.fromStopId, 0, false));
+            Coordinate c = firstSegment.geometry.getCoordinates()[0];
+            StopSpec stopSpec = new StopSpec(c.x, c.y);
+            if (firstSegment.fromStopId != null) {
+                stopSpec = new StopSpec(firstSegment.fromStopId);
+            }
+
+            Integer specificDwellTime = dwellTimes[realStopIndex++];
```

Comment:
If dwellTimes is set to an empty array above, won't this trigger an ArrayIndexOutOfBounds Exception rather than returning null?